File: app/src/main/java/ch/unstable/ost/theme/ThemeHelper.java
package ch.unstable.ost.theme;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.res.Resources;
import android.content.res.TypedArray;
import android.preference.PreferenceManager;
import android.support.annotation.AttrRes;
import android.support.annotation.DrawableRes;
import android.support.annotation.StyleRes;
import android.util.Log;
import android.util.TypedValue;

import ch.unstable.ost.R;
import ch.unstable.ost.preference.PreferenceKeys;

public enum ThemeHelper {
    ;

    private static final String TAG = "ThemeHelper";

    @StyleRes
    private static final int DEFAULT_THEME = R.style.GreenDarkTheme;

    public static void setTheme(Context context) {
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
        String currentTheme = preferences.getString(PreferenceKeys.KEY_THEME, null);
        int style;
        if (currentTheme == null) {
            style = DEFAULT_THEME;
        } else {
            style = getThemeStyle(context.getResources(), currentTheme);
        }
        context.setTheme(style);
    }

    @StyleRes
    private static int getThemeStyle(Resources resources, String currentTheme) {
        if (currentTheme == null) {
            throw new NullPointerException("currentTheme is null");
        }
        TypedArray styles = resources.obtainTypedArray(R.array.theme_styles);
        String[] values = resources.getStringArray(R.array.theme_values);

        int i = 0;
        int styleRes = 0;
        for (String value : values) {
            if (value.equals(currentTheme)) {
                styleRes = styles.getResourceId(i, 0);
                break;
            }
            ++i;
        }
        styles.recycle();
        if (styleRes == 0) {
            Log.e(TAG, "style not found (currentTheme: " + currentTheme + ")");
            return DEFAULT_THEME;
        }
        return styleRes;
    }

    @DrawableRes
    public static int getThemedDrawable(Context context, @AttrRes int attr) {
        Resources.Theme theme = context.getTheme();
        TypedValue typedValue = new TypedValue();
        theme.resolveAttribute(attr, typedValue, true);
        return typedValue.resourceId;
    }

    public static abstract class OnThemeChangeListener implements SharedPreferences.OnSharedPreferenceChangeListener {
        abstract void onThemeChanged();

        @Override
        public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
            if (key.equals(PreferenceKeys.KEY_THEME)) {
                onThemeChanged();
            }
        }
    }
}

File: app/src/main/java/ch/unstable/ost/theme/ThemeHelper.kt
package ch.unstable.ost.theme

import android.content.Context
import android.content.SharedPreferences
import android.content.res.Resources
import android.preference.PreferenceManager
import android.util.Log
import android.util.TypedValue
import androidx.annotation.AttrRes
import androidx.annotation.DrawableRes
import androidx.annotation.StyleRes
import ch.unstable.ost.R
import ch.unstable.ost.preference.PreferenceKeys

object ThemeHelper {

    abstract class OnThemeChangeListener : SharedPreferences.OnSharedPreferenceChangeListener {
        abstract fun onThemeChanged()
        override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {
            if (key == PreferenceKeys.KEY_THEME) {
                onThemeChanged()
            }
        }
    }

    private const val TAG = "ThemeHelper"

    @StyleRes
    private val DEFAULT_THEME = R.style.GreenDarkTheme
    fun setTheme(context: Context) {
        val preferences = PreferenceManager.getDefaultSharedPreferences(context)
        val currentTheme = preferences.getString(PreferenceKeys.KEY_THEME, null)
        val style: Int = if (currentTheme == null) {
            DEFAULT_THEME
        } else {
            getThemeStyle(context.resources, currentTheme)
        }
        context.setTheme(style)
    }

    @StyleRes
    private fun getThemeStyle(resources: Resources, currentTheme: String?): Int {
        if (currentTheme == null) {
            throw NullPointerException("currentTheme is null")
        }
        val styles = resources.obtainTypedArray(R.array.theme_styles)
        val values = resources.getStringArray(R.array.theme_values)
        var i = 0
        var styleRes = 0
        for (value in values) {
            if (value == currentTheme) {
                styleRes = styles.getResourceId(i, 0)
                break
            }
            ++i
        }
        styles.recycle()
        if (styleRes == 0) {
            Log.e(TAG, "style not found (currentTheme: $currentTheme)")
            return DEFAULT_THEME
        }
        return styleRes
    }

    @DrawableRes
    fun getThemedDrawable(context: Context, @AttrRes attr: Int): Int {
        val theme = context.theme
        val typedValue = TypedValue()
        theme.resolveAttribute(attr, typedValue, true)
        return typedValue.resourceId
    }
}